### PR TITLE
gh-119180: Avoid going through AST and eval() when possible in annotationlib

### PR DESCRIPTION
**Pull request by JelleZijlstra**
Often, ForwardRefs represent a single simple name. In that case, we can avoid going through the overhead of creating AST nodes and code objects and calling eval(): we can simply look up the name directly in the relevant namespaces.

This PR incorporates the changes from #124326 (which fixes a bug that I discovered while working on this change).

* Issue: gh-119180

### CODE REVIEW

File: Lib/test/test_annotationlib.py [hunk 1]
```diff
@@ -305,6 +311,23 @@ def test_fwdref_value_is_cached(self):
...
+            collections.Counter[int],
+        )
+
+    def test_name_lookup_without_eval(self):
+        # test the codepath where we look up simple names directly in the
+        # namespaces without going through eval()
+        self.assertIs(ForwardRef("int").evaluate(), int)
+        self.assertIs(ForwardRef("int").evaluate(locals={"int": str}), str)
+        self.assertIs(ForwardRef("int").evaluate(locals={"int": float}, globals={"int": str}), float)
+        self.assertIs(ForwardRef("int").evaluate(globals={"int": str}), str)
```
Comment (vstinner) [inline Lib/test/test_annotationlib.py:328]: You may also test mocking `builtins.int`. Something like:

```
with support.swap_attr(annotationlib.builtins, 'int', dict):
    self.assertIs(ForwardRef("int").evaluate(), dict)
```

Comment (JelleZijlstra) [reply Lib/test/test_annotationlib.py:328]: Good call, added

File: Lib/test/test_annotationlib.py [hunk 2]
```diff
@@ -280,7 +281,13 @@ class Gen[T]:
 
     def test_fwdref_with_module(self):
         self.assertIs(ForwardRef("Format", module="annotationlib").evaluate(), Format)
-        self.assertIs(ForwardRef("Counter", module="collections").evaluate(), collections.Counter)
+        self.assertIs(
+            ForwardRef("Counter", module="collections").evaluate(), collections.Counter
```
Comment (vstinner) [inline Lib/test/test_annotationlib.py]:
```suggestion
            ForwardRef("Counter", module="collections").evaluate(),
            collections.Counter
```

File: Lib/annotationlib.py [hunk 3]
```diff
@@ -267,18 +281,25 @@ def __init__(self, node, globals=None, owner=None, is_class=False, cell=None):
...
+                upper=self.__convert_to_ast(other.stop) if other.stop is not None else None,
+                step=self.__convert_to_ast(other.step) if other.step is not None else None,
             )
         else:
             return ast.Constant(value=other)
 
+    def __get_ast(self):
+        if isinstance(self.__ast_node__, str):
+            return ast.Name(id=self.__ast_node__)
+        return self.__ast_node__
```
Comment (vstinner) [inline Lib/annotationlib.py]: You might add a variable to not have to get/reference the attribute 3 times:

```suggestion
        node = self.__ast_node__
        if isinstance(node, str):
            return ast.Name(id=node)
        return node
```

**Review by vstinner — APPROVED** [review]
LGTM. I just have minor coding style suggestions that you can ignore if you prefer.

**Comment by JelleZijlstra** [comment]
Thanks @vstinner for the review!

File: Lib/annotationlib.py [hunk 4]
```diff
@@ -154,8 +155,19 @@ def evaluate(self, *, globals=None, locals=None, type_params=None, owner=None):
                     globals[param_name] = param
                     locals.pop(param_name, None)
 
-        code = self.__forward_code__
-        value = eval(code, globals=globals, locals=locals)
+        arg = self.__forward_arg__
+        if arg.isidentifier():
+            if arg in locals:
+                value = locals[arg]
+            elif globals is not None and arg in globals:
```
Comment (AlexWaygood) [inline Lib/annotationlib.py]: I don't think `globals` can be `None` here -- we make sure it's a `dict` no matter what on line 129 above

File: Lib/annotationlib.py [hunk 5]
```diff
@@ -154,8 +155,19 @@ def evaluate(self, *, globals=None, locals=None, type_params=None, owner=None):
                     globals[param_name] = param
                     locals.pop(param_name, None)
 
-        code = self.__forward_code__
-        value = eval(code, globals=globals, locals=locals)
+        arg = self.__forward_arg__
+        if arg.isidentifier():
```
Comment (AlexWaygood) [inline Lib/annotationlib.py]: Do you also need to check `not keyword.iskeyword(arg)`? (whether yes or no, I think you should add some tests with keywords :-)

**Review by AlexWaygood — APPROVED** [review]

**Pull request merged**